src/main/java/io/fabric8/elasticsearch/plugin/OpenShiftElasticSearchService.java

```diff
@@ -23,82 +23,93 @@
 
 import org.apache.logging.log4j.Logger;
 import org.elasticsearch.ElasticsearchException;
-import org.elasticsearch.client.Client;
+import org.elasticsearch.cluster.LocalNodeMasterListener;
 import org.elasticsearch.common.component.AbstractLifecycleComponent;
 import org.elasticsearch.common.logging.Loggers;
-import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.util.concurrent.EsExecutors;
 import org.elasticsearch.common.util.concurrent.FutureUtils;
 
-import io.fabric8.elasticsearch.plugin.acl.UserProjectCache;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import io.fabric8.elasticsearch.plugin.acl.ACLDocumentManager;
 
 /**
- * Service to handle spawning threads, lifecycles
+ * Service to handle spawning threads, lifecycles, and REST filter registrations
+ * 
  */
 public class OpenShiftElasticSearchService extends AbstractLifecycleComponent
-        implements ConfigurationSettings {
+        implements ConfigurationSettings, LocalNodeMasterListener {
 
-    private final Logger logger;
-    private final UserProjectCache cache;
-    private final Settings settings;
+    private static final Logger LOGGER = Loggers.getLogger(OpenShiftElasticSearchService.class);
+    
+    private final PluginSettings settings;
+    private final ACLDocumentManager aclManager;
     private ScheduledThreadPoolExecutor scheduler;
 
     @SuppressWarnings("rawtypes")
     private ScheduledFuture scheduledFuture;
 
-    // @TODO - Determine if we should be using the Threadpool provided by Elastic
-    public OpenShiftElasticSearchService(final Settings settings, final Client esClient, final UserProjectCache cache) {
-        super(settings);
-        this.settings = settings;
-        this.logger = Loggers.getLogger(getClass(), settings);
-        this.cache = cache;
+    public OpenShiftElasticSearchService( 
+            final ACLDocumentManager aclManager, 
+            final PluginSettings pluginSettings) {
+        super(pluginSettings.getSettings());
+        this.settings = pluginSettings;
+        this.aclManager = aclManager;
     }
-
+    
     @Override
     protected void doStart() throws ElasticsearchException {
+        LOGGER.debug("Started");
+    }
 
-        boolean dynamicEnabled = settings.getAsBoolean(OPENSHIFT_DYNAMIC_ENABLED_FLAG,
-                OPENSHIFT_DYNAMIC_ENABLED_DEFAULT);
-        logger.debug("Starting with Dynamic ACL feature enabled: {}", dynamicEnabled);
-
+    @Override
+    public void onMaster() {
```

Comment:
when does `onMaster` and `offMaster` trigger?

Reply:
It's a callback from Elasticsearch when a node is elected master or drops out